### PR TITLE
Playback Rate UI

### PR DESCRIPTION
**Pull request by jw-kaurand**
### What does this Pull Request do?
This PR adds in a UI for selecting playback rates for providers which support this option.

### Why is this Pull Request needed?
This is needed because it exposes the playback rate API functionality to the user via a menu.  It allows selection of either custom rates defined by the publisher to a set of standard rates available when setting up the player with it as an option.

### Are there any points in the code the reviewer needs to double check?
I have some concerns about the naming conventions involved with the SelectionDisplayMenu, especially its name.  The selection display menu may also be too narrowly tailored to the desired functionality, so perhaps looking at it to make it more simple or generic could help.  

### Are there any Pull Requests open in other repos which need to be merged with this?
There will be one in commercial to match this.  It will also have some tests related to this functionality.

#### Addresses Issue(s):

JW7-4038

### CODE REVIEW

File: src/js/api/config.js [hunk 1]
```diff
@@ -105,6 +109,33 @@ define([
             config.skin = config.skin.replace('.xml', '');
         }
 
+        if (config.playbackRateControls) {
+            var playbackRateOptions;
+
+            var validRatesFilter = (value) => { return _.isNumber(value) && value >= 0.25 && value <= 4; };
+
+            if (_.isBoolean(config.playbackRateControls)) {
+                playbackRateOptions = [0.5, 1, 1.25, 1.5, 2];
```
Comment (egreaves) [inline src/js/api/config.js]: Initialize `playbackRateOptions` to the defaults. That way, you only need to check `_.isArray(config.playbackRateControls)`.

File: src/js/api/config.js [hunk 2]
```diff
@@ -105,6 +109,33 @@ define([
...
+            } else if (_.isArray(config.playbackRateControls)) {
+                var filteredResults = config.playbackRateControls.filter(validRatesFilter);
+                playbackRateOptions = filteredResults.length ? filteredResults : false;
+            }
+
+            if (playbackRateOptions) {
```
Comment (egreaves) [inline src/js/api/config.js]: `playbackRateOptions` should be defined if `config.playbackRateControls` is set, so you don't need these conditions. I also don't understand why you'd reassign `config.playbackRateControls` AND set `config.playbackRateOptions`.

**Comment by egreaves** [comment]
@jw-kaurand Either mutate playbackRateControls or use playbackRateOptions (if you want to keep track of what was in the config) for what goes into the menu, but there isn't a need for both to end up with the same array of options.

**Comment by jw-kaurand** [comment]
A build of this in conjunction with the Commercial version of this can be found here: http://player-pr-test-jenkins.longtailvideo.com/builds/7425/archive/test/squash/

**Comment by jw-kaurand** [comment]
@egreaves `playbackRateControls` and `playbackRateOptions` aren't necessarily always going to be the same. `playbackRateOptions` is the value that is used to populate the playback rate menu, so it will potentially fluctuate between playlist items which do not support playback rate (LIVE streams, unsupported providers, etc) while `playbackRateControls` should always be what we reset `playbackRateOptions` back to when in situations where playback rate is supported.

**Comment by egreaves** [comment]
@jw-kaurand still seems like they're sharing responsibility. Use one property for keeping track of the menu and something else to determine if the menu should be visible or not based on the stream and provider.

File: src/js/controller/model.js [hunk 3]
```diff
@@ -236,6 +234,7 @@ define([
 
             // Attempt setting the playback rate to be the user selected value
             this.setPlaybackRate(this.get('defaultPlaybackRate'));
+            this.setPlaybackRatesAvailable();
```
Comment (egreaves) [inline src/js/controller/model.js]: The view should control whether the playbackRate controls are shown, not the model.

File: src/js/controller/model.js [hunk 4]
```diff
@@ -417,6 +417,15 @@ define([
             }
         };
 
+        this.setPlaybackRatesAvailable = function() {
+            var supportsPlaybackRates =
+                this.get('playbackRateControls') &&
+                _provider.supportsPlaybackRate &&
+                this.get('streamType') !== 'LIVE';
+
+            this.set('playbackRatesAvailable', supportsPlaybackRates);
```
Comment (egreaves) [inline src/js/controller/model.js]: This should be done in the view and doesn't require another model property.

File: src/js/view/controls/controlbar.js [hunk 5]
```diff
@@ -145,6 +149,11 @@ define([
                 hd: menu('jw-icon-hd', this._localization.hd),
                 cc: menu('jw-icon-cc', this._localization.cc),
                 audiotracks: menu('jw-icon-audio-tracks', this._localization.audioTracks),
+                playbackrates: new SelectionDisplayMenu(
+                    'jw-icon-playback-rate',
+                    this._localization.playbackRates,
```
Comment (egreaves) [inline src/js/view/controls/controlbar.js]: Is this the right property name?

Comment (jw-kaurand) [inline src/js/view/controls/controlbar.js]: Yup.  Matches the localization key.

**Comment by jw-kaurand** [comment]
The most recent build with the commercial version should be build number 7457.

**Comment by jw-kaurand** [comment]
test this please

**Comment by jw-kaurand** [comment]
The most recent build of this branch with the related commercial version should be available in build number 7506.

**Pull request merged**